beacon_node/beacon_chain/src/test_utils.rs

```diff
@@ -771,3 +650,436 @@ where
+    pub adversarial_validator_count: usize,
+    pub validators_keypairs: Vec<Keypair>,
+
+    pub chain: BeaconChain<T>,
+    pub spec: ChainSpec,
+    pub data_dir: TempDir,
+
+    pub rng: StdRng,
+}
+
+type TestRigAttestations<E> = Vec<(
+    Vec<(Attestation<E>, SubnetId)>,
+    Option<SignedAggregateAndProof<E>>,
+)>;
+
+impl<E: EthSpec> BeaconChainTestingRig<TestingRigType<E>> {
+    pub fn new(
+        eth_spec_instance: E,
+        honest_validator_count: usize,
+        adversarial_validator_count: usize,
+    ) -> Self {
+        let data_dir = tempdir().unwrap();
+        let mut spec = E::default_spec();
+
+        // Setting the target aggregators to really high means that _all_ validators in the
+        // committee are required to produce an aggregate. This is overkill, however with small
+        // validator counts it's the only way to be certain there is _at least one_ aggregator per
+        // committee.
+        spec.target_aggregators_per_committee = 1 << 32;
+
+        let validator_count = honest_validator_count + adversarial_validator_count;
+        let validators_keypairs =
+            types::test_utils::generate_deterministic_keypairs(validator_count);
+
+        let decorator = slog_term::PlainDecorator::new(slog_term::TestStdoutWriter);
+        let drain = slog_term::FullFormat::new(decorator).build();
+        let debug_level = slog::LevelFilter::new(drain, slog::Level::Debug);
+        let log = slog::Logger::root(std::sync::Mutex::new(debug_level).fuse(), o!());
+
+        let config = StoreConfig::default();
+        let store = Arc::new(HotColdDB::open_ephemeral(config, spec.clone(), log.clone()).unwrap());
+
+        let chain = BeaconChainBuilder::new(eth_spec_instance)
+            .logger(log.clone())
+            .custom_spec(spec.clone())
+            .store(store.clone())
+            .store_migrator(BlockingMigrator::new(store, log.clone()))
+            .data_dir(data_dir.path().to_path_buf())
+            .genesis_state(
+                interop_genesis_state::<E>(&validators_keypairs, HARNESS_GENESIS_TIME, &spec)
+                    .unwrap(),
+            )
+            .unwrap()
+            .dummy_eth1_backend()
+            .unwrap()
+            .null_event_handler()
+            .testing_slot_clock(HARNESS_SLOT_TIME)
+            .unwrap()
+            .build()
+            .unwrap();
+
+        // Nondeterminism in tests is a highly undesirable thing.  Seed the RNG to some arbitrary
+        // but fixed value for reproducibility.
+        let rng = StdRng::seed_from_u64(0x0DDB1A5E5BAD5EEDu64);
+
+        Self {
+            spec: chain.spec.clone(),
+            chain,
+            validators_keypairs,
+            data_dir,
+            honest_validator_count,
+            adversarial_validator_count,
+            rng,
+        }
+    }
+
+    pub fn slots_per_epoch(&self) -> u64 {
+        E::slots_per_epoch()
+    }
+
+    pub fn checkpoint_of_epoch(&self, epoch: u64) -> u64 {
```

Comment:
I think this func should be called something like `epoch_start_slot`, as checkpoint is already a defined term meaning a pair of `root` and `epoch`.

Reply:
Again, the paper says "We define checkpoint blocks to be blocks whose height is a multiple of a constant H (in
[7], H = 100)", but I'll rename to `epoch_start_slot`.

Reply:
Cool. I'd say the paper should be a secondary source to the spec (https://github.com/ethereum/eth2.0-specs/blob/v0.12.2/specs/phase0/beacon-chain.md#checkpoint)

Reply:
Definitely 👍